037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,251 @@
+spec:
+  # NOTE: mode and brokers fields are not available as in the KafkaRebalance,
+  #       they are automatically set on the corresponding KafkaRebalance by the operator
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The user has the flexibility to specify the [goals and the other rebalancing related configurations](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaRebalance-reference) the same way they already do today on a `KafkaRebalance` custom resource.
+The cluster operator will use that configuration to fill the corresponding `KafkaRebalance` but adding the `spec.mode` and `spec.brokers` accordingly.
+It would also contain a `spec.autoApproval` field, as described by this [proposal](https://github.com/strimzi/proposals/pull/56).
+
+We would also need to extend the `Kafka` custom resource to allow the user specifying that he wants auto-rebalancing on scaling by using a template.
+A new `spec.kafka.autoRebalance` field could be added for this purpose.
+Because we are talking about auto-rebalancing, the corresponding field could also be not limited to the rebalancing on scaling but a more general full continuous” auto-rebalancing to enable on the cluster.
+
+The `spec.kafka.autoRebalance` field allows to specify a list of auto-rebalancing modes with the corresponding templates to be used.
+To specify when running the auto-rebalancing, a `spec.kafka.autoRebalance.mode` field could be used to specify if running auto-rebalancing only on adding or removing brokers, or a full rebalance is requested.
+Having a full auto-rebalancing is out of scope for this proposal, the field is added for enabling a future implementation.
+
+The template to be used is referenced by a new mandatory `spec.kafka.autoRebalance.template` field and the cluster operator uses that template to create a corresponding `KafkaRebalance` custom resource.
+If the referenced `KafkaRebalanceTemplate` custom resource doesn't exist, the cluster operator will fail the reconcile.
+
+In the following example, a user can decide to have different rebalancing templates when adding or removing brokers and when they want a full rebalancing during the normal cluster execution.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      # using the same rebalancing template when adding or removing brokers
+      - mode: add-brokers
+        template: my-add-remove-brokers-rebalancing-template
+      - mode: remove-brokers
+        template: my-add-remove-brokers-rebalancing-template  
+      # using a different template for a full rebalancing  
+      - mode: full
+        template: my-full-rebalancing-template
```

Comment:
Should the `template` fields follow ObjectReference, TypedLocalObjectReference or LocalObjectReference APIs from Kubernetes?

Reply:
I'm inclined to this too, that way we're not implicitly coupling the `Kafka` API to the `KafkaRebalance` API. I.e. we'd leave the door open to a `BetterKafkaRebalance` kind in the future.

Reply:
Agree, I think `LocalObjectReference` would be enough specifying just the name. I will change it.